### PR TITLE
Suggest adding a where-clause when that can help

### PR DESCRIPTION
**Pull request by arielb1**
Suggest adding a where-clause when there is an unmet trait-bound that can be satisfied if some type can implement it.

r? @nikomatsakis

### CODE REVIEW

File: src/librustc/traits/error_reporting.rs [hunk 1]
```diff
@@ -357,14 +361,20 @@ pub fn report_selection_error<'a, 'tcx>(infcx: &InferCtxt<'a, 'tcx>,
...
                                 infcx.tcx.sess, obligation.cause.span, E0277,
-                                "the trait `{}` is not implemented for the type `{}`",
-                                trait_ref, trait_ref.self_ty());
-
-                            // Check if it has a custom "#[rustc_on_unimplemented]"
-                            // error message, report with that message if it does
-                            let custom_note = report_on_unimplemented(infcx, &trait_ref.0,
-                                                                      obligation.cause.span);
-                            if let Some(s) = custom_note {
+                                "the predicate `{}` is not satisfied",
```
Comment (nikomatsakis) [inline src/librustc/traits/error_reporting.rs]: Hmm, I wonder if it is better to write in this uniform style, or to try to have a custom error message? Not clear to me. It seems likely that if we are going to make this message generic, we could combine with some other error-code -- though maybe there is value in separating out this form of predicate from the others. NM, keep it with its own code.

File: src/librustc/traits/error_reporting.rs [hunk 2]
```diff
@@ -357,14 +361,20 @@ pub fn report_selection_error<'a, 'tcx>(infcx: &InferCtxt<'a, 'tcx>,
...
+                                trait_ref.to_predicate());
+
+                            // Try to report a good error message.
```
Comment (nikomatsakis) [inline src/librustc/traits/error_reporting.rs]: It'd be nice to be a bit more specific than this comment. For example, put in the body of each if:

`if ... {
    // Prefer to suggest an actionable fix `"add `T: Foo`"
} else {
    // Otherwise, custom explanation if available
}
```

At first, I was not sure if this should be an "xor" sort of logic -- but now I agree with you. So long as there is a chance that this is just a missing bound, it seems wrong to give a message like "you can't index a vector with an `X`" or whatever, since it may well be that you *could*. Might be worth adding (in the comment) your reasoning as to why it makes sense to give preference the actionable suggestion. (I am guessing as to your motivation here, maybe I have it wrong).

Basically these kind of "soft logic" decisions are really easy to lose track of, so leaving a detailed comment trail let's us refine them in the future without fear of overlooking use cases we have forgotten about.

File: src/test/compile-fail/trait-suggest-where-clause.rs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+use std::mem;
+
+struct Misc<T:?Sized>(T);
+
+fn check<T: Iterator, U: ?Sized>() {
+    // suggest a where-clause, if needed
+    mem::size_of::<U>();
+    //~^ ERROR `U : std::marker::Sized` is not satisfied
+    //~| HELP E0277
+    //~| HELP consider adding a `where U : std::marker::Sized` bound
```
Comment (nikomatsakis) [inline src/test/compile-fail/trait-suggest-where-clause.rs]: I wonder if we should leave the word `where` out from here. Not sure. Probably it's ok.

File: src/test/compile-fail/trait-suggest-where-clause.rs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+// except according to those terms.
+
+use std::mem;
+
+struct Misc<T:?Sized>(T);
+
+fn check<T: Iterator, U: ?Sized>() {
+    // suggest a where-clause, if needed
+    mem::size_of::<U>();
+    //~^ ERROR `U : std::marker::Sized` is not satisfied
```
Comment (nikomatsakis) [inline src/test/compile-fail/trait-suggest-where-clause.rs]: Given that I think the preferred format is `U: Sized`, we should adjust this print-out I suspect.

File: src/doc/book/closures.md [hunk 5]
```diff
@@ -371,13 +371,13 @@ assert_eq!(6, answer);
 This gives us these long, related errors:
 
 ```text
-error: the predicate `core::ops::Fn(i32) -> i32 : core::marker::Sized` is not satisfied [E0277]
+error: the trait bound `core::ops::Fn(i32) -> i32 : core::marker::Sized` is not satisfied [E0277]
```
Comment (nikomatsakis) [inline src/doc/book/closures.md]: ah, good

**Comment by nikomatsakis** [comment]
Everything looks good, I'd like to review the comments though, just because I'd like to ensure this is well documented for future reference (and to make sure we are on same page)

File: src/test/compile-fail/trait-suggest-where-clause.rs (hunk 3 above)
Comment (arielb1) [inline src/test/compile-fail/trait-suggest-where-clause.rs]: Maybe "consider adding a bound"? The bikeshed here is endless.

File: src/test/compile-fail/trait-suggest-where-clause.rs (hunk 4 above)
Comment (arielb1) [inline src/test/compile-fail/trait-suggest-where-clause.rs]: That's the `impl Display for TraitRef`.

**Comment by nikomatsakis** [comment]
@arielb1 

> That's the impl Display for TraitRef.

Yes, I know, I was suggesting changing it to remove the space.
